core/arch/riscv/kernel/thread_arch.c

```diff
@@ -0,0 +1,628 @@
+{
+	trap_regs->a0 = scall_regs->a0;
+	trap_regs->a1 = scall_regs->a1;
+	trap_regs->a2 = scall_regs->a2;
+	trap_regs->a3 = scall_regs->a3;
+	trap_regs->a4 = scall_regs->a4;
+	trap_regs->a5 = scall_regs->a5;
+	trap_regs->a6 = scall_regs->a6;
+	trap_regs->a7 = scall_regs->a7;
+	trap_regs->t0 = scall_regs->t0;
+	trap_regs->t1 = scall_regs->t1;
+}
+
+static void copy_trap_to_scall(struct thread_trap_regs *trap_regs,
+			       struct thread_scall_regs *scall_regs)
+{
+	*scall_regs = (struct thread_scall_regs) {
+		.status = trap_regs->status,
+		.ra = trap_regs->ra,
+		.a0 = trap_regs->a0,
+		.a1 = trap_regs->a1,
+		.a2 = trap_regs->a2,
+		.a3 = trap_regs->a3,
+		.a4 = trap_regs->a4,
+		.a5 = trap_regs->a5,
+		.a6 = trap_regs->a6,
+		.a7 = trap_regs->a7,
+		.t0 = trap_regs->t0,
+		.t1 = trap_regs->t1,
+	};
+}
+
+static void thread_user_ecall_handler(struct thread_trap_regs *trap_regs)
+{
+	struct thread_scall_regs scall_regs;
+
+	copy_trap_to_scall(trap_regs, &scall_regs);
+	thread_scall_handler(&scall_regs);
+	copy_scall_to_trap(&scall_regs, trap_regs);
+	trap_regs->epc += 4;
+}
+
+static void thread_exception_handler(unsigned long cause,
+				     struct thread_trap_regs *regs)
+{
+	switch (cause) {
+	case CAUSE_MISALIGNED_FETCH:
+	case CAUSE_FETCH_ACCESS:
+	case CAUSE_ILLEGAL_INSTRUCTION:
+	case CAUSE_BREAKPOINT:
+		DMSG("debug");
+		dump_trap_regs(regs);
+		break;
+	case CAUSE_MISALIGNED_LOAD:
+	case CAUSE_LOAD_ACCESS:
+	case CAUSE_MISALIGNED_STORE:
+	case CAUSE_STORE_ACCESS:
+	case CAUSE_SUPERVISOR_ECALL:
+	case CAUSE_VIRTUAL_SUPERVISOR_ECALL:
+	case CAUSE_MACHINE_ECALL:
+	case CAUSE_FETCH_PAGE_FAULT:
+	case CAUSE_LOAD_PAGE_FAULT:
+	case CAUSE_STORE_PAGE_FAULT:
+	case CAUSE_FETCH_GUEST_PAGE_FAULT:
+	case CAUSE_LOAD_GUEST_PAGE_FAULT:
+	case CAUSE_VIRTUAL_INSTRUCTION:
+	case CAUSE_STORE_GUEST_PAGE_FAULT:
+		thread_unhandled_trap(regs, cause);
+		break;
+	case CAUSE_USER_ECALL:
+		thread_user_ecall_handler(regs);
+		break;
+	default:
+		thread_unhandled_trap(regs, cause);
+		break;
+	}
+}
+
+static void thread_irq_handler(void)
+{
+	itr_core_handler();
+}
+
+static void thread_interrupt_handler(unsigned long cause,
+				     struct thread_trap_regs *regs)
+{
+	switch (cause) {
+	case IRQ_XTIMER:
+		clear_csr(CSR_XIE, CSR_XIE_TIE);
+		break;
+	case IRQ_XSOFT:
+		thread_unhandled_trap(regs, cause);
+		break;
+	case IRQ_XEXT:
+		thread_irq_handler();
+		break;
+	default:
+		thread_unhandled_trap(regs, cause);
+	}
+}
+
+void thread_trap_handler(long cause, unsigned long epc __maybe_unused,
+			 struct thread_trap_regs *regs,
+			 bool user __maybe_unused)
+{
+	/*
+	 * The Interrupt bit (XLEN-1) in the cause register is set
+	 * if the trap was caused by an interrupt.
+	 */
+	if (cause < 0)
+		thread_interrupt_handler(cause, regs);
+	/*
+	 * Otherwise, cause is never written by the implementation,
+	 * though it may be explicitly written by software.
+	 */
+	else
+		thread_exception_handler((cause & LONG_MAX), regs);
```

Comment:
Why do you mask it with `LONG_MAX`? Surely it's already positive and can't be larger than `LONG_MAX`.